src/puppetlabs/puppetdb/query_eng/engine.clj

```diff
@@ -240,59 +275,77 @@
                  "latest_report_noop" {:type :boolean
                                        :queryable? true
-                                       :field :reports.noop}
+                                       :field :reports.noop
+                                       :join-deps #{:reports}}
                  "latest_report_noop_pending" {:type :boolean
                                                :queryable? true
-                                               :field :reports.noop_pending}
+                                               :field :reports.noop_pending
+                                               :join-deps #{:reports}}
                  "latest_report_status" {:type :string
                                          :queryable? true
-                                         :field :report_statuses.status}
+                                         :field :report_statuses.status
+                                         :join-deps #{:report_statuses}}
                  "latest_report_corrective_change" {:type :boolean
                                                     :queryable? true
-                                                    :field :reports.corrective_change}
+                                                    :field :reports.corrective_change
+                                                    :join-deps #{:reports}}
                  "latest_report_job_id" {:type :string
                                          :queryable? true
-                                         :field :reports.job_id}
+                                         :field :reports.job_id
+                                         :join-deps #{:reports}}
                  "cached_catalog_status" {:type :string
                                           :queryable? true
-                                          :field :reports.cached_catalog_status}
+                                          :field :reports.cached_catalog_status
+                                          :join-deps #{:reports}}
                  "catalog_environment" {:type :string
                                         :queryable? true
-                                        :field :catalog_environment.environment}
+                                        :field :catalog_environment.environment
+                                        :join-deps #{:catalog_environment}}
                  "report_environment" {:type :string
                                        :queryable? true
-                                       :field :reports_environment.environment}}
+                                       :field :reports_environment.environment
+                                       :join-deps #{:reports_environment}}}
 
    :relationships certname-relations
 
    :selection {:from [:certnames]
+               ;; The join names here must match the values in
+               ;; :join-deps above, i.e. :foo or [:foo :bar].
```

Comment:
I believe there may be an issue with join-deps taking out some of the tables that are needed by other joins here. For example if you have a query like: 

`["from", "nodes",
             ["extract", ["report_timestamp", "catalog_environment"]]]`

This will cause the `:catalogs` table to be dropped out of the `:left-join` because the two extracted fields only require the `:report` and `[:environments :catalog_environment]` tables as far as join-deps is concerned. The `:environments` table aliased as `:catalog-environment`  requires the `:catalogs` table via `[:= :catalog_environment.id :catalogs.environment_id]` but `:catalogs` appears to be getting dropped out by join-deps if we extract fields that don't have a direct dependency on it. 

When I ran the query above I was getting the following error:

`avax.servlet.ServletException: org.postgresql.util.PSQLException: ERROR: missing FROM-clause entry for table "catalogs"`

Maybe we could add `:catalogs` as an additional join-dep to the `catalog_environment` field and look for any other places where there could be similar dependencies between tables.

Reply:
Excellent catch -- I'm fixing these up, and adding tests for the relevant cases.